Technical review: Add WebAuthn Signal API (#37557)
mdn/content: files/en-us/web/api/publickeycredential/signalallacceptedcredentials_static/index.md

---
title: "PublicKeyCredential: signalAllAcceptedCredentials() static method"
short-title: signalAllAcceptedCredentials()
slug: Web/API/PublicKeyCredential/signalAllAcceptedCredentials_static
page-type: web-api-static-method
status:
  - experimental
browser-compat: api.PublicKeyCredential.signalAllAcceptedCredentials_static
---

{{APIRef("Web Authentication API")}}{{securecontext_header}}{{seecompattable}}

The **`signalAllAcceptedCredentials()`** static method of the {{domxref("PublicKeyCredential")}} interface signals to the authenticator all of the valid [credential IDs](/en-US/docs/Web/API/PublicKeyCredentialRequestOptions#id) that the [relying party](https://en.wikipedia.org/wiki/Relying_party) server still holds for a particular user.

> ## Description

Reviewer: The API template puts Description below Syntax/before Examples, so this should be moved.

Note, I don't particularly agree with this rule, but consistency is important.

Author: OK, good call. In each case, I have moved the Description section to the proper position and added a clarifying 2nd paragraph to the intro.

### `signalAllAcceptedCredentials()` versus `signalUnknownCredential()`

It may seem like `signalAllAcceptedCredentials()` and {{domxref("PublicKeyCredential.signalUnknownCredential_static", "signalUnknownCredential()")}} have similar purposes, so what situation should each one be used in?

To be clear, `signalAllAcceptedCredentials()` should only be used in cases where the user is authenticated and you want to update the state of a user's credentials. Don't use it in cases where the user is not authenticated — for example if authentication failed because the user deleted a credential from the relying party but later tried to sign in with it because it was still showing in the authenticator. Using `signalAllAcceptedCredentials()` for this purpose would share the entire list of `credentialId`s for a given user with an unauthenticated party, which may not be desirable.

> In cases where the user is not authenticated, you should instead use `signalUnknownCredential()`. It doesn't reveal the user's `credentialId` list to the authenticator, minimizing the data shared with an unauthenticated party.

Reviewer: See https://github.com/mdn/content/pull/37557/files#r1927988031

I don't think this is the right place for **all** this information - i.e. I don't think there is a confusion here about the methods. I would capture most of this in the intro.

Author: As it stands now, we are providing a very brief description in the intro, and a more detailed description here. The Google folks wanted it to be made clear when you should use each method. I think it is useful;  not sure if it just needs to be cut down a bit?

Reviewer: Sure, but you can make that clear without the for comparison here. As per https://github.com/mdn/content/pull/37557/files#r1931317437 a comparison is useful , but it should go at the **top** level web API doc and here we should just make the rules for this method clear. It is more important here than in the other case though, because here it does matter if we call the method when not authenticated.

See suggestion https://github.com/mdn/content/pull/37557/files#r1931325003 and https://github.com/mdn/content/pull/37557/files#r1931381906

Author: I've dealt with your other comments, but I'll deal with the comparison thing later on.

## Syntax

```js-nolint
signalAllAcceptedCredentials(options)
```

### Parameters

- `options`
  - : An object representing the valid credentials, which contains the following properties:
    - `allAcceptedCredentialIds`
      - : An array of base64url-encoded strings representing the [`id`s of the credentials](/en-US/docs/Web/API/PublicKeyCredentialRequestOptions#id) that are still valid.
    - `rpId`
      - : A string representing the [`id` of the relying party](/en-US/docs/Web/API/PublicKeyCredentialCreationOptions#id_2) that sent the signal.
    - `userId`
      - : A base64url-encoded string representing the [`id` of the user](/en-US/docs/Web/API/PublicKeyCredentialCreationOptions#id_3) the credentials relate to.

### Return value

A {{jsxref("Promise")}} that resolves to {{jsxref("undefined")}}.

### Exceptions

The promise rejects with a `TypeError` {{domxref("DOMException")}} if the `userId` or any of the `allAcceptedCredentialIds` elements are not valid base64url-encoded strings.

## Examples

> 

Reviewer: We normally now always have subheadings for all examples.

Something like

```
### Signaling the accepted credentials
```

Author: added, in all three cases

> The following snippet shows typical usage of `signalAllAcceptedCredentials()`:

Reviewer: This example is OK. But I think the examples in the explainer that do things like "The authenticator has AA and BB, and then you specify this signal AA, will remove BB are more complete. Perhaps worth a few more words than this?

Author: I've just deleted the "typical usage" lines.

I don't understand this stuff well enough to create a detailed "real" example. I'm just trying to get it documented. If you really think these examples are a red flag, we'll have to go back to the Google folks and try to get more information.

Reviewer: > If you really think these examples are a red flag,

I don't.  This is a good fix for me, though I have suggested https://github.com/mdn/content/pull/37557/files#r1931387840

```js

> // Feature detection

Reviewer: Why this line?

Author: Well, it is feature detection...

I've deleted those lines from the code examples.

---
title: "PublicKeyCredential: signalAllAcceptedCredentials() static method"
short-title: signalAllAcceptedCredentials()
slug: Web/API/PublicKeyCredential/signalAllAcceptedCredentials_static
page-type: web-api-static-method
status:
  - experimental
browser-compat: api.PublicKeyCredential.signalAllAcceptedCredentials_static
---

{{APIRef("Web Authentication API")}}{{securecontext_header}}{{seecompattable}}

The **`signalAllAcceptedCredentials()`** static method of the {{domxref("PublicKeyCredential")}} interface signals to the authenticator all of the valid [credential IDs](/en-US/docs/Web/API/PublicKeyCredentialRequestOptions#id) that the [relying party](https://en.wikipedia.org/wiki/Relying_party) (RP) server still holds for a particular user.

> This allows the authenticator to update credential information and remove credentials that are no longer regognized by the RP, such as those for deleted accounts. It should only be used when the current user is authenticated — after sign in, and following deletion of a credential on the RP site.

Reviewer: Typo, plus perhaps minor tidy in case there are other reasons than deletion for doing this (and in any case, I would expand on the fact that it exposes sensitive info more.

```suggestion
This allows the authenticator to update credential information and/or remove credentials that are no longer recognized by the RP, such as those for deleted accounts. It should only be called when the current user is authenticated — after sign up or sign-in  —  as it exposes sensitive information belonging to the user. 
```

Author: This is good, and I've used it, with a couple of tweaks.

## Syntax

```js-nolint
signalAllAcceptedCredentials(options)
```

### Parameters

- `options`
  - : An object representing the valid credentials, which contains the following properties:
    - `allAcceptedCredentialIds`
      - : An array of base64url-encoded strings representing the [`id`s of the credentials](/en-US/docs/Web/API/PublicKeyCredentialRequestOptions#id) that are still valid.
    - `rpId`
      - : A string representing the [`id` of the RP](/en-US/docs/Web/API/PublicKeyCredentialCreationOptions#id_2) that sent the signal.
    - `userId`
      - : A base64url-encoded string representing the [`id` of the user](/en-US/docs/Web/API/PublicKeyCredentialCreationOptions#id_3) the credentials relate to.

### Return value

A {{jsxref("Promise")}} that resolves to {{jsxref("undefined")}}.

### Exceptions

> The promise rejects with a `TypeError` {{domxref("DOMException")}} if the `userId` or any of the `allAcceptedCredentialIds` elements are not valid base64url-encoded strings.

Reviewer: See https://github.com/mdn/content/pull/37557/files#r1931308584

Author: updated

It also rejects with a `SecurityError` {{domxref("DOMException")}} if the RP domain is not valid.

> ## Description
> 
> It is possible for the information stored in a user's authenticator about a [discoverable credential](/en-US/docs/Web/API/Web_Authentication_API#discoverable_credentials_and_conditional_mediation) (for example, [a passkey](https://passkeys.dev/)) to go out sync with the server. This usually occurs when the user deletes a credential from the RP site without updating the authenticator.
> 
> The next time they try to sign in with a discoverable credential, the deleted credential will still be presented in the relevant UI, but the sign-in attempt will fail because the RP server won't recognize it. This results in a confusing user experience.
> 
> To avoid this issue, `signalAllAcceptedCredentials()` should be called on the RP site each time a user deletes a credential or signs in, to tell the authenticator which credentials are still valid for the given user. It is up to the authenticator how to handle this information, but the expectation is that it will synchronize its information with the provided credentials list. Credentials that don't appear in the list should be removed so that the user won't be offered credentials that don't exist in the sign-in UI.
> 
> > [!WARNING]
> > Exercise caution when invoking `signalAllAcceptedCredentials()` — any valid credentials not included in the list are intended to be removed from the authenticator, which will prevent the user from signing in with them. Passing an empty list may remove all of the user's credentials. Some authenticators may support restoring credentials via a subsequent call to `signalAllAcceptedCredentials()` with the previously removed credential IDs included in the list.
> 
> ### `signalAllAcceptedCredentials()` versus `signalUnknownCredential()`
> 
> It may seem like `signalAllAcceptedCredentials()` and {{domxref("PublicKeyCredential.signalUnknownCredential_static", "signalUnknownCredential()")}} have similar purposes, so what situation should each one be used in?
> 
> To be clear, `signalAllAcceptedCredentials()` should only be used in cases where the user is authenticated and you want to update the state of a user's credentials. Don't use it in cases where the user is not authenticated — for example if authentication failed because the user deleted a credential from the RP but later tried to sign in with it because it was still showing in the authenticator. Using `signalAllAcceptedCredentials()` for this purpose would share the entire list of `credentialId`s for a given user with an unauthenticated party, which may not be desirable.
> 
> In cases where the user is not authenticated, you should instead use `signalUnknownCredential()`. It doesn't reveal the user's `credentialId` list to the authenticator, minimizing the data shared with an unauthenticated party.

Reviewer: - signalAllAcceptedCredentials is called by the web app "on the authenticator" not "on the RP site." - i.e. its an update to the authenticator, not the backend, as indicated.
- This cleans up potentially multiple credentials pre-emptively - i.e. you've already authenticated so it won't be the credential that you just used that is out of sync, unless you just deleted it. Upshot though, we use plural here.
- The method mitigates, but does not avoid the issue.
- In this page we do need to provide some comparative info, because this can't be called anytime. But most should be on the main page.

```suggestion
## Description

It is possible for the information stored in a user's authenticator about [discoverable credentials](/en-US/docs/Web/API/Web_Authentication_API#discoverable_credentials_and_conditional_mediation) (for example, [a passkey](https://passkeys.dev/)) to go out of sync with the RP server. This usually occurs when the user deletes credentials from the RP site without updating the authenticator.

When a user attempts to sign-in, the browser presents the discoverable credentials stored on the authenticator.
If the user selects a credential that was removed from the RP server, it won't be recognized, and the sign-in attempt 
will fail. This is a confusing user experience!

To mitigate this issue, once a user is authenticated, we can call `signalAllAcceptedCredentials()` to tell the authenticator which credentials are valid for the given user. 
It is up to the authenticator how to handle this information, but the expectation is that it will synchronize its information with the provided credentials list. Credentials that don't appear in the list should be removed so that the user won't be offered credentials that don't exist in the sign-in UI.
Note that the method exposes user information, which is why it should only be called when the user is authenticated!

`signalAllAcceptedCredentials()` should be called immediately after a user authenticates (signs-in or up).
The method should also be called if an authenticated user deletes a credential.

> [!WARNING]
> Exercise caution when invoking `signalAllAcceptedCredentials()` — any valid credentials not included in the list are intended to be removed from the authenticator, which will prevent the user from signing in with them. Passing an empty list may remove all of the user's credentials. Some authenticators may support restoring credentials via a subsequent call to `signalAllAcceptedCredentials()` with the previously removed credential IDs included in the list.

> [!NOTE]
> It is possible for stored information to be deleted when a user is not authenticated, in which case a user might be presented with an invalid credential.
> You shouldn't call `signalAllAcceptedCredentials()` in this case because it exposes user information.
> Instead you call {{domxref("PublicKeyCredential.signalUnknownCredential_static", "signalUnknownCredential()")}} with the specific credential to be deleted.
> For more information see [THE COMPARATIVE OVERVIEW I HOPE YOU WILL ADD].
```

Author: I meant "called by the RP site" — it is still the RP site that calls the method, even though it is the authenticator that is signaled to update its details.

I have made some tweaks to fix your issues, but I'm keeping the text largely the same for now. It'll probably change again when I try to add the comparison you want me to add.

## Examples

### Signaling the accepted credentials

> 

Reviewer: Perhaps

```suggestion

Given an authenticated user, the following code can be used to update the authenticator with the credentials that will be accepted by the RP server.

```

Author: I added a sentence

    - `rpId`
      - : A string representing the [`id` of the RP](/en-US/docs/Web/API/PublicKeyCredentialCreationOptions#id_2) that sent the signal.
    - `userId`
      - : A base64url-encoded string representing the [`id` of the user](/en-US/docs/Web/API/PublicKeyCredentialCreationOptions#id_3) the credentials relate to.

### Return value

A {{jsxref("Promise")}} that resolves to {{jsxref("undefined")}}.

### Exceptions

The promise rejects with the following exceptions:

- `SecurityError` {{domxref("DOMException")}}
  - : The RP domain is not valid.
- `TypeError` {{domxref("DOMException")}}
  - : The `userId` or any of the `allAcceptedCredentialIds` elements are not valid base64url-encoded strings.

## Description

It is possible for the information stored in a user's authenticator about a [discoverable credential](/en-US/docs/Web/API/Web_Authentication_API#discoverable_credentials_and_conditional_mediation) (for example, [a passkey](https://passkeys.dev/)) to go out sync with the server. This usually occurs when the user deletes a credential from the RP site without updating the authenticator.

> The next time they try to sign in with a discoverable credential, the deleted credential will still be presented in the relevant UI, but the sign-in attempt will fail because the RP server won't recognize it. This results in a confusing user experience.

Reviewer: Flow not quite right - they could sign in with a valid credential even if there are deleted credentials in the set they are presented with - same issue as https://github.com/mdn/content/pull/37557/files#r1936492650

```suggestion
When a user attempts to login using discoverable credentials, they are presented with a set of credentials from the authenticator, and the selected credential is returned to the RP web app.
If the user selects a credential that has been deleted from the RP server, it won't be recognized, and the login will fail.
This is a confusing experience for users, who except to only be offered credentials that should succeed.
```

Author: updated with the same text as on the `signalUnknownCredential()` page.

> To mitigate this issue, `signalAllAcceptedCredentials()` should be called by the RP site each time a user deletes a credential or signs in, to tell the authenticator which credentials are still valid for the given user. It is up to the authenticator how to handle this information, but the expectation is that it will synchronize its information with the provided credentials list. Credentials that don't appear in the list should be removed so that the user won't be offered credentials that don't exist in the sign-in UI.

Reviewer: FWIW and just thinking aloud, I wonder if there is a way when we write "called by the RP site" to more clearly indicate that this is code running in the browser. Probably not

Author: This makes sense. In my next commit, I've gone through all the updated pages in this PR and made this change where appropriate.

> [!WARNING]
> Exercise caution when invoking `signalAllAcceptedCredentials()` — any valid credentials not included in the list are intended to be removed from the authenticator, which will prevent the user from signing in with them. Passing an empty list may remove all of the user's credentials. Some authenticators may support restoring credentials via a subsequent call to `signalAllAcceptedCredentials()` with the previously removed credential IDs included in the list.

`signalAllAcceptedCredentials()` should _only_ be called when the current user is authenticated because it exposes sensitive information belonging to the user. If the user is not authenticated because they tried to log in with a credential that does not exist on the RP server, you should instead call {{domxref("PublicKeyCredential.signalUnknownCredential_static", "PublicKeyCredential.signalUnknownCredential()")}} with the unrecognized credential, so the authenticator can delete it. See [Discoverable credential synchronization methods](/en-US/docs/Web/API/Web_Authentication_API#discoverable_credential_synchronization_methods) for a more detailed comparison.

## Examples

### Signaling the accepted credentials

> In this example, we invoke the `signalAllAcceptedCredentials()` method, passing it the details of a credential the user has just logged in with. As a result, the authenticator should update its own copy of the credentials so that it stays in sync with the RP.

Reviewer: Not quite. While you have in this case passed in the details of the credential the user has just logged in with, actually you are passing in all credentials.

```suggestion
In this example, we invoke the `signalAllAcceptedCredentials()` method, passing it the details of all credentials belonging to the user, including those they just logged in with. As a result, the authenticator should update its own copy of the credentials so that they stays in sync with the RP.
```

Author: Thank you. Updated in next commit.

> [!WARNING]
> Exercise caution when invoking `signalAllAcceptedCredentials()` — any valid credentials not included in the list are intended to be removed from the authenticator, which will prevent the user from signing in with them. Passing an empty list may remove all of the user's credentials. Some authenticators may support restoring credentials via a subsequent call to `signalAllAcceptedCredentials()` with the previously removed credential IDs included in the list.

### `signalAllAcceptedCredentials()` versus `signalUnknownCredential()`

It may seem like `signalAllAcceptedCredentials()` and {{domxref("PublicKeyCredential.signalUnknownCredential_static", "PublicKeyCredential.signalUnknownCredential()")}} have similar purposes, so which one should be used when?

> To be clear, `signalAllAcceptedCredentials()` should only be used in cases where authentication is _successful_ and you want to update the state of a user's credentials. Don't use it in cases where authentication failed, for example because the user deleted a credential from the relying party but later tried to sign in with it because it was still showing in the authenticator. Using `signalAllAcceptedCredentials()` for this purpose would share the entire list of `credentialId`s for a given user with an unauthenticated party, which may not be desirable.

Reviewer 2: Optional, for clarity:

```suggestion
To be clear, `signalAllAcceptedCredentials()` should only be used in cases where the user is authenticated and you want to update the state of a user's credentials. Don't use it in cases where the user is not authenticated, for example because the user deleted a credential from the relying party but later tried to sign in with it because it was still showing in the authenticator. Using `signalAllAcceptedCredentials()` for this purpose would share the entire list of `credentialId`s for a given user with an unauthenticated party, which may not be desirable.
```

> In cases where authentication _failed_, you should instead use `signalUnknownCredential()`. It only passes a single `credentialId` to the authenticator, minimizing the data shared with an unauthenticated party.

Reviewer 2: ```suggestion
In cases where the user is not authenticated, you should instead use `signalUnknownCredential()`. It doesn't reveal the list of credential IDs to the authenticator, minimizing the data shared with an unauthenticated party.
```

`signalUnknownCredential` doesn't reveal any of the credentials, since the credential ID in the method is the same the client just tried to use.
